File: 125-valid-palindrome/valid-palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
     if (s.length === 0) return true;

    const isAlphaNumeric = (char) => /[a-zA-Z0-9]/.test(char);

    let left = 0, right = s.length - 1;

    while (left < right) {
        if (!isAlphaNumeric(s[left])) left++;
        else if (!isAlphaNumeric(s[right])) right--;
        else if (s[left++].toLowerCase() !== s[right--].toLowerCase()) return false;
    }

    return true;
}

File: 125-valid-palindrome/valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if not s:
            return True

        def is_alpha_numeric(char):
            return char.isalnum()

        left, right = 0, len(s) - 1

        while left < right:
            if not is_alpha_numeric(s[left]):
                left += 1
            elif not is_alpha_numeric(s[right]):
                right -= 1
            elif s[left].lower() != s[right].lower():
                return False
            else:
                left += 1
                right -= 1

        return True
